### DISCUSSION TITLE
chacha6 instead?

### DISCUSSION THREAD

**Discussion by boss1819** [database64128/swgp-go#40] [General]
since encryption is already handled by wireguard, did you think about using chacha6 or chacha8 instead, as suggested by Wireguard founder to a similar project (xt_wgobfs) here
https://lists.zx2c4.com/pipermail/wireguard/2022-September/007831.html

that would likely lead to better performance without affecting the obfuscation.
what do you think?

**Comment by database64128** [comment]
> a similar project (xt_wgobfs)

Not so similar. xt_wgobfs is a kernel module, whereas swgp-go is a userspace proxy utility.

> that would likely lead to better performance without affecting the obfuscation.

It won't. swgp-go's throughput bottleneck is the syscall interface for sending and receiving packets, not its cryptography (at least in zero-overhead mode).

In our zero-overhead (obfuscation) mode, we only read random bytes from the system (`getrandom(2)`) for the XChaCha20-Poly1305 nonce when handling handshake packets. Normally this only occurs every 2 minutes, so there's no point in switching to a weaker PRNG. Data packets only need a simple AES operation on a single block, which is usually hardware-accelerated on modern CPUs.

Our paranoid mode uses AEAD on whole packets. It's basically a generic UDP tunnel. While it could probably benefit from some performance optimizations, I'm not sure if it's worth the added complexities. If you want fast generic proxies (that's also a lot more complex in its implementation), use https://github.com/database64128/shadowsocks-go instead.

**Comment by boss1819** [comment]
thanks for the clarifications.
I am using the paranoid mode and it is circumventing the block. I will give it a try also with the zero-overhead mode, to see if still successful.
if yes, will use it instead then since it seems from your explanation should have superior performance.
